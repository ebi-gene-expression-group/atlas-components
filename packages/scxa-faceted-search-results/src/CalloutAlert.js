import React from 'react'
import PropTypes from 'prop-types'

const CalloutAlert = ({error}) =>
  <div className={`row column`}>
    <div className={`callout alert small`}>
      <h5>Oops!</h5>
      <p>
        {error.description}<br/>
        If the error persists, in order to help us debug the issue, please copy the URL and this message and
        send it to us via <a href={`https://www.ebi.ac.uk/support/gxasc`}>the EBI Support & Feedback system</a>:
      </p>
      <code>{`${error.name}: ${error.message}`}</code>
    </div>
  </div>

CalloutAlert.propTypes = {
  error: PropTypes.shape({
    description: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
  })
}

export default CalloutAlert
