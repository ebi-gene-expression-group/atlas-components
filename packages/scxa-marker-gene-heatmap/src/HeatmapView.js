import React from 'react'
import PropTypes from 'prop-types'
import URI from 'urijs'
import MarkerGeneHeatmap from './MarkerGeneHeatmap'
import PlotSettingsDropdown from './PlotSettingsDropdown'
import LoadingOverlay from './LoadingOverlay'

const CalloutAlert = ({error}) =>
  <div className={`row column`}>
    <div className={`callout alert small`}>
      <h5>Oops!</h5>
      <p>
        {error.description}<br/>
        If the error persists, in order to help us debug the issue, please copy the URL and this message and
        send it to us via <a href={`https://www.ebi.ac.uk/support/gxasc`}>the EBI Support & Feedback system</a>:
      </p>
      <code>{`${error.name}: ${error.message}`}</code>
    </div>
  </div>

CalloutAlert.propTypes = {
  error: PropTypes.shape({
    description: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
  })
}

class HeatmapView extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      data: [],
      isLoading: true,
      hasError: null
    }
  }

  async _fetchAndSetState({resource, host}) {
    this.setState({
      isLoading: true
    })

    const url = URI(resource, host).toString()

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      this.setState({
        data: await response.json(),
        isLoading: false,
        hasError: null
      })
    } catch(e) {
      this.setState({
        data: null,
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message
        }
      })
    }
  }

  componentDidMount() {
    this._fetchAndSetState(this.props)
  }

  componentDidUpdate(previousProps) {
    if (previousProps.resource !== this.props.resource) {
      this._fetchAndSetState(this.props)
    }
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: {
        description: `There was a problem rendering this component.`,
        name: error.name,
        message: `${error.message} â€“ ${info}`
      }
    })
  }

  render() {
    const { data, isLoading, hasError } = this.state

    const { ks, selectedK, onSelectK, wrapperClassName, plotWrapperClassName, heatmapHeight } = this.props

    const kOptions = ks.sort((a, b) => a-b).map((k) => ({
      value: k.toString(),
      label: `k = ${k}`
    }))


    return (
      hasError ?
        <CalloutAlert error={hasError}/> :
        <div className={wrapperClassName}>
          <div className={plotWrapperClassName}>
            <PlotSettingsDropdown
              key={`selectK`}
              labelText={`View marker genes for:`}
              options={kOptions}
              onSelect={(selectedOption) => {onSelectK(selectedOption.value)}}
              defaultValue={{value: selectedK, label: `k = ${selectedK}`}}
            />
            <div key={`heatmap`} style={{position: `relative`}} className={wrapperClassName}>
              <MarkerGeneHeatmap
                key={`heatmap`}
                data={data}
                chartHeight={heatmapHeight}
              />
              <LoadingOverlay
                show={isLoading}
              />
            </div>
          </div>
        </div>
    )
  }
}

HeatmapView.propTypes = {
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  ks: PropTypes.arrayOf(PropTypes.number).isRequired,
  selectedK: PropTypes.number.isRequired,
  onSelectK: PropTypes.func,
  wrapperClassName: PropTypes.string,
  plotWrapperClassName: PropTypes.string,
  heatmapHeight: PropTypes.number
}

HeatmapView.defaultProps = {
  wrapperClassName: `row`,
  plotWrapperClassName: `medium-12 columns`,
  heatmapHeight: 800
}

export default HeatmapView
