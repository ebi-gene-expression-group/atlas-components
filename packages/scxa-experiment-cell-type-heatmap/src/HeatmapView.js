import React from 'react'
import PropTypes from 'prop-types'
import LoadingOverlay from './LoadingOverlay'
import CalloutAlert from './CalloutAlert'
import { cellTypeHeatmapView } from './cellTypeHeatmapView'
import URI from 'urijs'
import _ from "lodash";

class HeatmapView extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      data: [],
      filteredData: [],
      selectedClusterId: null,
      isLoading: true,
      hasError: null
    }
    this.onSelectCluster = this.onSelectCluster.bind(this)
  }

  async _fetchAndSetState({resource, host}) {
    this.setState({
      isLoading: true
    })

    const url = URI(resource, host).toString()
    console.log(`url`, url)
    try {
      const response = await fetch(url)
      console.log(`response`, response)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      const jsonData = await response.json()
      this.setState({
        data: await jsonData,
        filteredData: await jsonData,
        selectedClusterId: null,
        isLoading: false,
        hasError: null
      })
    } catch(e) {
      console.log(`catch`)

      this.setState({
        data: null,
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message
        }
      })
    }
  }

  componentDidMount() {
    this._fetchAndSetState(this.props)
  }

  componentDidUpdate(previousProps) {
    if (previousProps.resource !== this.props.resource) {
      this._fetchAndSetState(this.props)
    }
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: {
        description: `There was a problem rendering this component.`,
        name: error.name,
        message: `${error.message} â€“ ${info}`
      }
    })
  }

 onSelectCluster(selectedOption) {
   this.setState((state) => ({
     data: _.cloneDeep(state.data),
     filteredData: selectedOption.value === `all` ?
       _.cloneDeep(state.data) :
       _.filter(state.data, {'clusterIdWhereMarker': parseInt(selectedOption.value)}),
     selectedClusterId: selectedOption
   }))
 }

render() {
    const { isLoading, hasError } = this.state
    const { wrapperClassName, plotWrapperClassName } = this.props
    const view = cellTypeHeatmapView(this.props.props, this.state)

    return (
      hasError ?
        <CalloutAlert error={hasError}/> :
        <div>
          { this.props.type == `cluster` ? view[1] : `` }
          <div className={wrapperClassName}>
            <div className={plotWrapperClassName} style={{position: `relative`}}>
              {view[0]}
              <LoadingOverlay
                show={isLoading}
              />
            </div>
          </div>
        </div>
    )
  }
}

HeatmapView.propTypes = {
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  props: PropTypes.object.isRequired
}

export default HeatmapView
