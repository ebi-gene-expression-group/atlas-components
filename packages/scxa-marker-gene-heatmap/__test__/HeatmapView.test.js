import React from 'react'
import Enzyme from 'enzyme'
import {shallow} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import fetchMock from 'fetch-mock'

import HeatmapView from '../src/HeatmapView'
import CalloutAlert from '../src/CalloutAlert'

Enzyme.configure({ adapter: new Adapter() })

describe(`HeatmapView`, () => {
  beforeEach(() => {
    fetchMock.restore()
  })

  const props = {
    host: `foo/`,
    resource: `bar`,
    ks: [1, 2, 3, 4],
    selectedK: `1`
  }

  test(`Renders error if API request is unsuccessful`, () => {
    const wrapper = shallow(<HeatmapView {...props} />)
    expect(wrapper.exists(CalloutAlert)).toBe(true)
  })
})
