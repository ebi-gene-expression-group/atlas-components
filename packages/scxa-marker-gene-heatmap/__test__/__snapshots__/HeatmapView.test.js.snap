// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatmapView matches snapshot when heatmap type is celltypes 1`] = `
<HeatmapView
  defaultHeatmapHeight={300}
  hasDynamicHeight={true}
  heatmapRowHeight={20}
  heatmapType="celltypes"
  host="foo/"
  plotWrapperClassName=""
  resource="bar"
  species="species"
  wrapperClassName=""
>
  <CalloutAlert
    error={
      Object {
        "description": "There was a problem communicating with the server. Please try again later.",
        "message": "fetch is not defined",
        "name": "ReferenceError",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="columns large-9 large-centered"
      >
        <div
          className="callout alert small"
        >
          <h5>
            Oops!
          </h5>
          <p>
            There was a problem communicating with the server. Please try again later.
            <br />
            If the error persists, in order to help us debug the issue, please copy the URL and this message and send it to us via 
            <a
              href="https://www.ebi.ac.uk/support/gxasc"
            >
              the EBI Support & Feedback system
            </a>
            :
          </p>
          <code>
            ReferenceError: fetch is not defined
          </code>
        </div>
      </div>
    </div>
  </CalloutAlert>
</HeatmapView>
`;

exports[`HeatmapView matches snapshot when heatmap type is clusters 1`] = `
<HeatmapView
  defaultHeatmapHeight={300}
  hasDynamicHeight={true}
  heatmapRowHeight={20}
  heatmapType="clusters"
  host="foo/"
  plotWrapperClassName=""
  resource="bar"
  species="species"
  wrapperClassName=""
>
  <CalloutAlert
    error={
      Object {
        "description": "There was a problem communicating with the server. Please try again later.",
        "message": "fetch is not defined",
        "name": "ReferenceError",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="columns large-9 large-centered"
      >
        <div
          className="callout alert small"
        >
          <h5>
            Oops!
          </h5>
          <p>
            There was a problem communicating with the server. Please try again later.
            <br />
            If the error persists, in order to help us debug the issue, please copy the URL and this message and send it to us via 
            <a
              href="https://www.ebi.ac.uk/support/gxasc"
            >
              the EBI Support & Feedback system
            </a>
            :
          </p>
          <code>
            ReferenceError: fetch is not defined
          </code>
        </div>
      </div>
    </div>
  </CalloutAlert>
</HeatmapView>
`;

exports[`HeatmapView matches snapshot when heatmap type is multiexperimentcelltypes 1`] = `
<HeatmapView
  defaultHeatmapHeight={300}
  hasDynamicHeight={true}
  heatmapRowHeight={20}
  heatmapType="multiexperimentcelltypes"
  host="foo/"
  plotWrapperClassName=""
  resource="bar"
  species="species"
  wrapperClassName=""
>
  <CalloutAlert
    error={
      Object {
        "description": "There was a problem communicating with the server. Please try again later.",
        "message": "fetch is not defined",
        "name": "ReferenceError",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="columns large-9 large-centered"
      >
        <div
          className="callout alert small"
        >
          <h5>
            Oops!
          </h5>
          <p>
            There was a problem communicating with the server. Please try again later.
            <br />
            If the error persists, in order to help us debug the issue, please copy the URL and this message and send it to us via 
            <a
              href="https://www.ebi.ac.uk/support/gxasc"
            >
              the EBI Support & Feedback system
            </a>
            :
          </p>
          <code>
            ReferenceError: fetch is not defined
          </code>
        </div>
      </div>
    </div>
  </CalloutAlert>
</HeatmapView>
`;
